### DISCUSSION TITLE
Multiple Success and Failure Actions per Rule

### DISCUSSION THREAD

**Discussion by NicoMackLIST** [microsoft/RulesEngine#333] [Ideas]
Unless I'm mistaken, rules currently allow one single OnFailure and OnSuccess action. I'm currently implementing rules in the field of construction logistics, and I came across a few use cases where being able to have multiple OnFailure / OnSuccess actions would be very convenient. Has anyone else come across such use cases?

**Comment by nitingautam** [comment]
my understanding is you can extend functionality and can create custom actions as you need. Please have a look [https://github.com/microsoft/RulesEngine/wiki/Actions](https://github.com/microsoft/RulesEngine/wiki/Actions)

**Reply by pburrows** [reply]
Weirdly, the above linked URL doesn't go to the text of the URL (it navigates to https://github.com/microsoft/RulesEngine/discussions/url instead of that Wiki page)

**Comment by bmasotta-lcg** [comment]
Same here! For a fully generic rule usage It would be great to have multiple actions per rule result (don't forget to include an execution sequence enforcing mechanism).
If not, We will need to encode multiple action data in a single action context and then decode it in the CustomAction.

I think we can hack this limit by indirectly executing actions by using an OutputExpression Action that calls ad hoc Custom Type methods that returns a boolean value.

But this is only a hack and not the proper way of coding rules!


```
"Actions": {
  "OnSuccess": {
    "Name": "OutputExpression",
      "Context": {
      "Expression": "VariableContext.Write(PreviousAlert, true).True && Messenger.Send(PendingCommand).True"
    }
  }
}
```

VariableContext and Messenger are Custom Types
.Write and Send methods return a ad hoc class that allows to select the proper boolean value to prevent partial expression evaluation (short circuit)

**Reply by djl-85** [reply]
Hi there, very interesting hack. But, does it work?
